### PR TITLE
Support for building upstream LLVM from source

### PR DESCRIPTION
**Pull request by jgravelle-google**
@kripken @juj @dschuff

### CODE REVIEW

**Comment by jgravelle-google** [comment]
@juj @kripken ?
This is probably going to start being useful pretty quick.
We do also need to get an upstream LLVM build on the prebuilts, which I'm not sure how we do.

File: emsdk_manifest.json [hunk 1]
```diff
@@ -145,6 +145,34 @@
...
   },
+  {
+    "id": "upstream-clang",
+    "version": "master",
+    "bitness": 32,
+    "install_path": "clang/upstream",
+    "git_branch": "master",
+    "url": "https://github.com/llvm-mirror/llvm.git",
+    "clang_url": "https://github.com/llvm-mirror/clang.git",
+    "lld_url": "https://github.com/llvm-mirror/clang.git",
```
Comment (juj) [inline emsdk_manifest.json]: Should this be `https://github.com/llvm-mirror/lld.git`?

Comment (jgravelle-google) [reply emsdk_manifest.json]: Yes it should... why did I think this worked locally?
Oh wow I had the 64-bit version correct, so it happened to work on my machine.

File: emsdk [hunk 2]
```diff
@@ -821,7 +834,8 @@ def build_fastcomp_tool(tool):
 
   enable_assertions = ENABLE_LLVM_ASSERTIONS.lower() == 'on' or (ENABLE_LLVM_ASSERTIONS == 'auto' and build_type.lower() != 'release' and build_type.lower() != 'minsizerel')
 
-  args = ['-DLLVM_TARGETS_TO_BUILD=X86;JSBackend', '-DLLVM_INCLUDE_EXAMPLES=OFF', '-DCLANG_INCLUDE_EXAMPLES=OFF', '-DLLVM_INCLUDE_TESTS=' + tests_arg, '-DCLANG_INCLUDE_TESTS=' + tests_arg, '-DLLVM_ENABLE_ASSERTIONS=' + ('ON' if enable_assertions else 'OFF')]
+  js_target = ';JSBackend' if build_js else ''
```
Comment (juj) [inline emsdk]: How about this:

```python
only_supports_wasm = hasattr(tool, 'only_supports_wasm')

targets_to_build = 'X86'
if not only_supports_wasm: targets_to_build += ';JSBackend'
```

and then below:

```c++
if ENABLE_WASM or only_supports_wasm:
  args += ['-DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD=WebAssembly']
```

That would then allow deleting the function `build_upstream_llvm_tool()` altogether (`ENABLE_WASM` variable can be left in whatever user-specified state in that case), and `def build_llvm_tool(tool, build_js):` -> `def build_llvm_tool(tool):`

File: emsdk_manifest.json [hunk 3]
```diff
@@ -145,6 +145,34 @@
...
+    "bitness": 64,
+    "install_path": "clang/upstream",
+    "git_branch": "master",
+    "url": "https://github.com/llvm-mirror/llvm.git",
+    "clang_url": "https://github.com/llvm-mirror/clang.git",
+    "lld_url": "https://github.com/llvm-mirror/lld.git",
+    "custom_install_script": "build_llvm",
+    "activated_path": "%installation_dir%/%fastcomp_build_bin_dir%",
+    "activated_cfg": "LLVM_ROOT='%installation_dir%/%fastcomp_build_bin_dir%'",
+    "cmake_build_type": "Release"
```
Comment (juj) [inline emsdk_manifest.json]: and then add `only_supports_wasm: true` here?

File: emsdk (hunk 2 above)
Comment (jgravelle-google) [reply emsdk]: Sure, done

**Comment by juj** [comment]
Perfect, many thanks!

**Pull request merged**